File: c++/20 - Valid Parentheses.cpp
// 20
// Valid Parentheses
// https://leetcode.com/problems/valid-parentheses/
// c++
// easy
// O(n)
// O(n)
// string stack

// very fast solution using switch and stack
// O(n), O(n)
class Solution {
public:
    bool isValid(string s) {
        stack<char> stck;
        
        for (char& c : s) {
            switch (c) {
                case '[': stck.push(']'); break;
                case '{': stck.push('}'); break;
                case '(': stck.push(')'); break;
                default:
                    if (stck.empty() || c != stck.top()) return false;
                    stck.pop();
            }
        }
        
        return stck.empty();
    }
};

File: python/20 - Valid Parentheses.py
# 20
# Valid Parentheses
# https://leetcode.com/problems/valid-parentheses/
# python
# easy
# O(n)
# O(n)
# string stack

# Initial solution in python using python [O(n), O(n)]
# Impoved version done in c++
class Solution:
    def isValid(self, s: str) -> bool:
        d = {
            "[": "]",
            "{": "}",
            "(": ")"
        }
        
        stack = []
        
        for c in s:
            if stack and stack[-1] == c:
                del stack[-1]
                continue
                
            if c in d.keys():
                stack.append(d[c])
            elif c in d.values():
                return False    

        return len(stack) == 0
